File: Hackathon/2026/Steward/web/guardian/src/lib/heartbeat.ts
// The guardian's INDEPENDENT proof-of-life check — the verifying half of the trustless
// dead-man's-switch. Before a guardian will arm an InheritanceRelease it fetches the owner's
// signed heartbeat bulletin and re-verifies it HERE, on-device, then decides `is_lapsed`
// itself. It never trusts the coordinator's `state` flag: a relay with no key cannot forge a
// proof-of-life, and cannot move the owner's signed timestamp without breaking this check.
//
// THE canonical message — MUST be byte-identical to steward_core::heartbeat (Rust) and the
// owner console:
//   msg = "steward-heartbeat-v1" || vault_id (UTF-8) || unix_time (u64, 8 bytes BIG-ENDIAN)
// Proven byte-for-byte against ed25519-dalek in the task's cross-impl check.
import * as ed from '@noble/ed25519'

const DOMAIN = new TextEncoder().encode('steward-heartbeat-v1')

/** Build the canonical heartbeat message (see the module header). */
export function heartbeatMessage(vaultId: string, time: number): Uint8Array {
  const id = new TextEncoder().encode(vaultId)
  const msg = new Uint8Array(DOMAIN.length + id.length + 8)
  msg.set(DOMAIN, 0)
  msg.set(id, DOMAIN.length)
  // u64 BIG-ENDIAN unix time
  new DataView(msg.buffer).setBigUint64(DOMAIN.length + id.length, BigInt(time), false)
  return msg
}

function fromHex(h: string): Uint8Array {
  const clean = h.trim()
  if (clean.length % 2 !== 0 || /[^0-9a-fA-F]/.test(clean)) throw new Error('bad hex')
  const out = new Uint8Array(clean.length / 2)
  for (let i = 0; i < out.length; i++) out[i] = parseInt(clean.slice(i * 2, i * 2 + 2), 16)
  return out
}

/** Verify a signed heartbeat over the canonical message — the guardian doing it ITSELF, with
 *  `@noble/ed25519` (WebCrypto SHA-512). Returns `false` on any malformed input or failure;
 *  mirrors `steward_core::heartbeat::verify_heartbeat` exactly. */
export async function verifyHeartbeat(
  pubkeyHex: string,
  vaultId: string,
  time: number,
  sigHex: string,
): Promise<boolean> {
  try {
    const pk = fromHex(pubkeyHex)
    const sig = fromHex(sigHex)
    if (pk.length !== 32 || sig.length !== 64) return false
    return await ed.verifyAsync(sig, heartbeatMessage(vaultId, time), pk)
  } catch {
    return false
  }
}

/** Whether the switch has lapsed at `now`, judged from the owner's own signed timestamp:
 *  `now > latest_time + interval + grace`. Strict `>` — exactly at the deadline is NOT yet
 *  lapsed. Mirrors `steward_core::heartbeat::is_lapsed`. */
export function isLapsed(
  latestTime: number,
  intervalSecs: number,
  graceSecs: number,
  now: number,
): boolean {
  return now > latestTime + intervalSecs + graceSecs
}

File: Hackathon/2026/Steward/crates/steward-ffi/src/heartbeat.rs
//! Heartbeat — the guardian-verifiable proof-of-life (§6 of the spec).
//!
//! Thin UniFFI wrappers over [`steward_core::heartbeat`]. The canonical message is
//! `"steward-heartbeat-v1" ‖ vault_id ‖ time_be`, byte-identical to the Rust core,
//! the web apps (`@noble/ed25519`) and ZecAuth — so an owner, a guardian, and this
//! SDK all agree signature-for-signature.

use rand::rngs::OsRng;
use rand::RngCore;

use crate::error::StewardError;

/// A freshly generated Ed25519 heartbeat keypair, returned to the owner at vault
/// setup. Record the `pubkey_hex` on the vault; keep the `secret_hex` on-device and
/// use it to sign each proof-of-life.
#[derive(uniffi::Record)]
pub struct HeartbeatKey {
    /// 32-byte Ed25519 secret-key **seed**, hex (RFC 8032; never leaves the device).
    pub secret_hex: String,
    /// 32-byte Ed25519 public key, hex — publish this on the vault record.
    pub pubkey_hex: String,
}

/// Sign a proof-of-life for `vault_id` at `time` (unix seconds) with the owner's
/// hex heartbeat secret seed. Deterministic (RFC 8032): identical inputs → identical
/// bytes. Returns the 64-byte signature as hex.
#[uniffi::export]
pub fn sign_heartbeat(sk_hex: String, vault_id: String, time: u64) -> Result<String, StewardError> {
    Ok(steward_core::heartbeat::sign_heartbeat_hex(&sk_hex, &vault_id, time)?)
}

/// Verify a hex heartbeat signature against the vault's recorded hex public key and
/// the canonical message. Returns `false` — never throws — on any decode or
/// verification failure. A relay has no secret key, so it cannot forge a `true` here.
#[uniffi::export]
pub fn verify_heartbeat(pubkey_hex: String, vault_id: String, time: u64, sig_hex: String) -> bool {
    steward_core::heartbeat::verify_heartbeat_hex(&pubkey_hex, &vault_id, time, &sig_hex)
}

/// Derive the hex Ed25519 public key from a hex secret seed.
#[uniffi::export]
pub fn heartbeat_pubkey(sk_hex: String) -> Result<String, StewardError> {
    Ok(steward_core::heartbeat::public_key_hex(&sk_hex)?)
}

/// Whether the dead-man's-switch has **lapsed** at `now`, judged purely from the
/// owner's own signed timestamp: `now > latest + interval + grace` (saturating).
/// This is the authoritative inheritance gate a guardian computes on its own clock
/// from a heartbeat it verified itself — never trusting a relay's state flag.
#[uniffi::export]
pub fn is_lapsed(latest: u64, interval: u64, grace: u64, now: u64) -> bool {
    steward_core::heartbeat::is_lapsed(latest, interval, grace, now)
}

/// Generate a fresh Ed25519 heartbeat keypair (owner setup). The secret is 32
/// random bytes from the platform CSPRNG; the public key is derived from it.
#[uniffi::export]
pub fn generate_heartbeat_key() -> HeartbeatKey {
    let mut seed = [0u8; 32];
    OsRng.fill_bytes(&mut seed);
    // `public_key` accepts any 32-byte seed and cannot fail; derive without a
    // hex round-trip so the whole call is infallible.
    let pubkey_hex = hex::encode(steward_core::heartbeat::public_key(&seed));
    let secret_hex = hex::encode(seed);
    HeartbeatKey { secret_hex, pubkey_hex }
}
